Changed to using normal version of Mjolnir

zwall-bp opened the pull request:
There were issues providing the auto-reporting capabilities the customer was asking for, so I had to rewrite the Mjolnir resource to use the standard version, where it requires a user access token, instead of being deployed as an appservice.

Extra steps are needed in order to deploy Mjolnir now, but the complexity has been eased by the use of a script, `/root/create-account.sh`.

Commit:
Changed to using normal version of Mjolnir

ramereth commented on files/default/create-account.sh:
```diff
@@ -0,0 +1,23 @@
+set -e
+
+MJOLNIR_USERNAME=mjolnir
+MJOLNIR_PASSWD=$(dd if=/dev/urandom bs=16 count=1 2>/dev/null | base64 )
```
Can you please use `pwgen` instead?

ramereth commented on files/default/create-account.sh:
```diff
@@ -0,0 +1,23 @@
+
+echo Credentials have been saved to /root/mjolnir-user.txt
+
+rm -- $0
```
Please run `shellcheck` and fix the linting issues mentioned.

ramereth commented on resources/mjolnir.rb:
```diff
@@ -18,7 +18,8 @@
 property :key_appservice, String, default: lazy { osl_matrix_genkey(host_name + container_name) }
 property :key_homeserver, String, default: lazy { osl_matrix_genkey(container_name + host_name) }
-property :bot_name, String, default: 'Auto Mod'
+property :access_token, String, required: true
```
Should this be a sensitive property?

ramereth commented on resources/mjolnir.rb:
```diff
@@ -103,4 +79,12 @@
   end
+
+  # Generate Account Creation Script
+  cookbook_file '/root/create-account.sh' do
```
Please put this in `/usr/local/sbin` and call it `matrix-create-account`

ramereth commented on resources/mjolnir.rb:
```diff
@@ -103,4 +79,12 @@
+    source 'create-account.sh'
+    cookbook 'osl-matrix'
+    mode '0700'
+    not_if { ::File.exist?('/root/mjolnir-user.txt') }
```
I'm a little confused. Is it intended that this script is only needed once manually, or is it used in automation somehow?

ramereth commented on resources/synapse_service.rb:
```diff
@@ -9,6 +9,7 @@
 property :config_hookshot, Hash, default: {}
 property :config_matrix_irc, Hash, default: {}
 property :config_mjolnir, Hash, default: {}
+property :mjolnir_token, String
```
Should this be a sensitive property?

ramereth commented on resources/mjolnir.rb:
```diff
@@ -38,54 +39,29 @@
-        'volumes' => [
-          "#{new_resource.host_path}/appservice-data/mjolnir_init.sql:/docker-entrypoint-initdb.d/init.sql:ro",
-          "#{new_resource.host_path}/appservice-data/#{new_resource.container_name}-db:/var/lib/postgresql/data",
-        ],
```
Why was this removed?

zwall-bp commented:
@ramereth Answering all of the questions at once, Mjolnir is a bit different from other Matrix bots, as it is not normally an Appservice, which is a feature that allows for a limited set of features while being easier to develop and deploy bots to. Instead, Mjolnir requires the use of an actual user account, with admin privilages.

Previously, I stated that I set it up to use its appservice version, instead. But, unfortunately, it does not come with the feature requirement our customers at Inkscape wanted, automatically post Reports into the moderation chat. As such, I had to redesign the resource to instead run Mjolnir the normal way.

In order to create a user, and get their session token, the bash script I created is supposed to be a one-time-use, in order to sign-up, sign-in, and get the required credentials. This has to be done manually, unfortunately. The script is designed to self-delete, as running it a second time will fail due to the account already being created.

The normal verison of Mjolnir does not require a database in order to function, that is (for some reason) strictly a requirement of the appservice version.

I didn't use `pwgen` as our image doesn't come with it, stock, but I'll add it as a package.

ramereth commented:
It might be better to create a custom resource to handle that in a more chef-like way. I asked Co-pilot to give me something, and it gave me something like the following:

```ruby
# Custom resource for managing Matrix users
resource_name :matrix_user

property :username, String, name_property: true
property :password, String, required: true
property :admin, [true, false], default: false
property :homeserver_url, String, required: true
property :admin_token, String, required: true

action :create do
  converge_by("Create Matrix user #{new_resource.username}") do
    require 'net/http'
    require 'uri'
    require 'json'

    uri = URI("#{new_resource.homeserver_url}/_synapse/admin/v1/register")
    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = uri.scheme == 'https'

    request = Net::HTTP::Post.new(uri.path, {
      'Content-Type' => 'application/json',
      'Authorization' => "Bearer #{new_resource.admin_token}"
    })

    request.body = {
      username: new_resource.username,
      password: new_resource.password,
      admin: new_resource.admin
    }.to_json

    response = http.request(request)

    unless response.code.to_i == 200
      raise "Failed to create Matrix user: #{response.code} - #{response.body}"
    end

    Chef::Log.info("Matrix user #{new_resource.username} created successfully.")
  end
end
```

Example usage:
```ruby
matrix_user 'newuser' do
  password 'securepassword'
  admin false
  homeserver_url 'https://your-homeserver.com'
  admin_token 'YOUR_ADMIN_ACCESS_TOKEN'
  action :create
end
```

Obviously needs to be tweaked and cleaned up, but you might give that a try.

zwall-bp commented:
I'd like to add it to the resource, my only issue is a race condition of waiting for the server to be deployed. As the Mjolnir resource is not able to schedule `post` notifications at the end of a chef run. Only at the end of the resource's run. I could try and make it another action, which is then ran afterwards, but there's still the worry of the Synapse server deploying in time, in order to work.

zwall-bp commented:
Wait, then there will be a race condition between generating an account and populating Mjolnir's config...

ramereth commented:
@zwall-bp is there something you can query from the API to know if it's online or not?

Commit:
Created osl_matrix_user

zwall-bp commented:
@ramereth 

I've created a new resource for automating the process of creating a Matrix user within Chef, `osl_matrix_user`. It both registers, and acquires a session token, which can then be used for allowing Mjolnir to interact with the server.

The session token can  be found at `/root/mjolnir-access-token.txt`. There might be a way to fully automate the deployment of Mjolnir, by passing in that info into Mjolnir? There's a race condition there, in what resource should go first.

As of right now, Chef will sleep for 2 minutes after deploying the docker containers, in order to give enough time for `osl_matrix_user` to run, as it requires an active Synapse server. It might also be possible to create a polling resource, simply waiting for a `200` from `http://localhost8008`

zwall-bp commented:
The script created previously can also be removed, as it's no longer needed.

ramereth committed:
Major refactor

This refactors much of the logic and code globally to make it better to use and
automate.

- Remove unused account script
- Add new helper methods to handle server status, create/update users, register
  initial admin user, getting a user token via logging in, creating channels,
  and dynamically listing the app services via the yaml files generated
- Create new resources:
  - osl_matrix_admin_user
  - osl_matrix_channel
  - osl_matrix_user
- Split out synapse server into it's own docker compose environment to make it
  easier to handle bootstrapping issues
- Create unified docker network for all services
- Set the compose name to something easier to reach and manage based on the host
  domain property
- Mark additional resources as sensitive that were printing secrets
- Update properties in mjolnir resource to something better
- Add new properties to osl_synapse_service resource including a required
  admin_password property.
- Ensure we redeploy/restart synapse server
- Add more unit tests to ensure full coverage
- Setup Postgresql server with ala carte suite and ensure we test it properly
- Improve test coverage of integration tests

Signed-off-by: Lance Albertson <lance@osuosl.org>

ramereth commented:
!bump minor ~

osuosl-manatee commented:
Jenkins has merged this PR into `master` and has automatically performed a minor-level version bump to v1.7.0.  Have a nice day!
